pkg/concurrency/runner.go

```diff
@@ -48,12 +60,10 @@ func (r *RunnerManager) Add(runner Runner) error {
 // runner returns, the RunnerManager will stop all other runners and return any
 // error.
 func (r *RunnerManager) Run(ctx context.Context) error {
-	select {
-	case <-r.running:
-		return errors.New("manager already started")
-	default:
+	if r.running.Load() {
```

Comment:
Between `r.running.Load()` and `r.running.Store(true)` there can be a race condition, so you can have 2 goroutines continuing with the init at the same time.

The nice thing about sync.Atomic is the compare-and-swap which is fully atomic.

So:

```go
if !r.running.CompareAndSwap(false, true) {
Return ErrManagerAlreadyStarted
}
```

This works because CompareAndSwap atomically (in a single operation) checks if the value was false, and only if it was false, it sets it to true. If the swap happened (i.e. it was false before and now it’s true), it returns true. If the value was already true (Run was already called), then no swap happens.